File: amazon-scraper/python/productVariations.py
import urllib.parse
import csv
import re
import requests
from bs4 import BeautifulSoup

# Configuration
token = "<SDO-token>"
asin = "B0CG84XR6N"  # Change this to any product ASIN with variations
geocode = "us"
zipcode = "10001"

# Build product URL from ASIN
target_url = f"https://www.amazon.com/dp/{asin}"

# Initial page scrape to find dimensions
targetUrl = urllib.parse.quote(target_url)
apiUrl = "https://api.scrape.do/plugin/amazon/?token={}&url={}&geocode={}&zipcode={}&output=html".format(token, targetUrl, geocode, zipcode)
response = requests.request("GET", apiUrl)

soup = BeautifulSoup(response.text, "html.parser")

# Extract available dimensions and options
dimensions = {}
for row in soup.find_all("div", {"id": re.compile(r"inline-twister-row-.*")}):
    dim_name = row.get("id", "").replace("inline-twister-row-", "").replace("_name", "")
    options = []
    for option in row.find_all("li", {"data-asin": True}):
        opt_asin = option.get("data-asin")
        if opt_asin and option.get("data-initiallyUnavailable") != "true":
            swatch = option.find("span", {"class": "swatch-title-text-display"})
            img = option.find("img")
            button = option.find("span", {"class": "a-button-text"})

            option_name = swatch.text.strip() if swatch else img.get("alt", "").strip() if img and img.get("alt") else button.get_text(strip=True) if button and button.get_text(strip=True) != "Select" else "Unknown"
            options.append({"name": option_name, "asin": opt_asin})

    if options:
        dimensions[dim_name] = options

print(f"Found dimensions: {list(dimensions.keys())}")
for dim_name, options in dimensions.items():
    print(f"  {dim_name}: {len(options)} options")

# Setup dimension traversal order and CSV headers
priority = ['color', 'size', 'style', 'pattern', 'material', 'fit']
dim_names = sorted(dimensions.keys(), key=lambda x: priority.index(x.lower()) if x.lower() in priority else len(priority))
headers = ["ASIN", "Product Name"] + dim_names + ["Price"]
results = []
scraped = set()

# Recursive function to scrape all product variations
def scrape_variations(variation_asin, dim_index=0, prefix=""):
    if variation_asin in scraped:
        return

    url = f"https://www.amazon.com/dp/{variation_asin}/?th=1&psc=1"
    targetUrl = urllib.parse.quote(url)
    apiUrl = "https://api.scrape.do/plugin/amazon/?token={}&url={}&geocode={}&zipcode={}&output=html".format(token, targetUrl, geocode, zipcode)
    response = requests.request("GET", apiUrl)

    soup = BeautifulSoup(response.text, "html.parser")

    title_elem = soup.find(id="productTitle")
    if not title_elem:
        print(f"{prefix}[SKIP] {variation_asin}: Product page not available")
        scraped.add(variation_asin)
        return
    name = title_elem.text.strip()

    try:
        price = "Out of Stock" if soup.find("div", id="outOfStockBuyBox_feature_div") else f"${soup.find(class_='a-price-whole').text}{soup.find(class_='a-price-fraction').text}"
    except:
        price = "N/A"

    page_dims = {}
    for row in soup.find_all("div", {"id": re.compile(r"inline-twister-row-.*")}):
        dim_name = row.get("id", "").replace("inline-twister-row-", "").replace("_name", "")
        options = []
        for option in row.find_all("li", {"data-asin": True}):
            opt_asin = option.get("data-asin")
            if opt_asin and option.get("data-initiallyUnavailable") != "true":
                swatch = option.find("span", {"class": "swatch-title-text-display"})
                img = option.find("img")
                button = option.find("span", {"class": "a-button-text"})

                option_name = swatch.text.strip() if swatch else img.get("alt", "").strip() if img and img.get("alt") else button.get_text(strip=True) if button and button.get_text(strip=True) != "Select" else "Unknown"
                options.append({"name": option_name, "asin": opt_asin})

        if options:
            page_dims[dim_name] = options

    # End of recursion - collect final data
    if dim_index >= len(dim_names) or not page_dims:
        scraped.add(variation_asin)
        selections = {}
        for dim_name in dim_names:
            for row_id in [f"inline-twister-row-{dim_name}_name", f"inline-twister-row-{dim_name}"]:
                row = soup.find("div", {"id": row_id})
                if row:
                    selected = row.find("span", {"class": re.compile(r".*a-button-selected.*")})
                    if selected:
                        swatch = selected.find("span", {"class": "swatch-title-text-display"})
                        img = selected.find("img")
                        selections[dim_name] = swatch.text.strip() if swatch else img.get("alt", "").strip() if img and img.get("alt") else selected.get_text(strip=True) if "Select" not in selected.get_text(strip=True) else "N/A"
                        break
                    else:
                        for option in row.find_all("li", {"data-asin": variation_asin}):
                            swatch = option.find("span", {"class": "swatch-title-text-display"})
                            button = option.find("span", {"class": "a-button-text"})
                            selections[dim_name] = swatch.text.strip() if swatch else button.get_text(strip=True) if button and "Select" not in button.get_text(strip=True) else "N/A"
                            break
                    break

        row = [variation_asin, name] + [selections.get(dim, "N/A") for dim in dim_names] + [price]
        results.append(row)

        sel_str = ", ".join([f"{k}:{v}" for k, v in selections.items()])
        print(f"{prefix}[OK] {variation_asin}: {price} | {sel_str}")
        return

    # Continue recursion through dimensions
    current_dim = dim_names[dim_index]
    if current_dim in page_dims:
        options = page_dims[current_dim]
        if dim_index == 0:
            print(f"{prefix}Found {len(options)} {current_dim} options")

        for i, option in enumerate(options):
            if dim_index == 0:
                print(f"{prefix}{current_dim} {i+1}/{len(options)}: {option['name']}")
            scrape_variations(option["asin"], dim_index + 1, prefix + "    ")
    else:
        scrape_variations(variation_asin, dim_index + 1, prefix)

# Start scraping process
print(f"\nDimension traversal order: {dim_names}")
print(f"\nStarting variation crawling...")

scrape_variations(asin)

# Export to CSV
csv_file = "productVariations.csv"

with open(csv_file, 'w', newline='', encoding='utf-8') as f:
    writer = csv.writer(f)
    writer.writerow(headers)
    writer.writerows(results)

print(f"\nDone! Scraped {len(results)} unique variations")
print(f"Data exported to {csv_file}")

File: amazon-scraper/node.js/productVariations.js
const axios = require('axios');
const cheerio = require('cheerio');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Configuration
const token = "<SDO-token>";
const asin = "B0CG84XR6N"; // Change this to any product ASIN with variations
const geocode = "us";
const zipcode = "10001";

// Build product URL from ASIN
const targetUrl = `https://www.amazon.com/dp/${asin}`;

const scraped = new Set();
const results = [];
let dimNames = [];
let headers = [];

async function fetchPage(url) {
    const encodedUrl = encodeURIComponent(url);
    const apiUrl = `https://api.scrape.do/plugin/amazon/?token=${token}&url=${encodedUrl}&geocode=${geocode}&zipcode=${zipcode}&output=html`;
    const response = await axios.get(apiUrl);
    return cheerio.load(response.data);
}

function extractDimensions($) {
    const dimensions = {};
    $('div[id^="inline-twister-row-"]').each((i, row) => {
        const dimName = $(row).attr('id').replace('inline-twister-row-', '').replace('_name', '');
        const options = [];

        $(row).find('li[data-asin]').each((j, option) => {
            const optAsin = $(option).attr('data-asin');
            const isUnavailable = $(option).attr('data-initiallyunavailable') === 'true';

            if (optAsin && !isUnavailable) {
                const swatch = $(option).find('span.swatch-title-text-display');
                const img = $(option).find('img');
                const button = $(option).find('span.a-button-text');

                let optionName = 'Unknown';
                if (swatch.length) {
                    optionName = swatch.text().trim();
                } else if (img.length && img.attr('alt')) {
                    optionName = img.attr('alt').trim();
                } else if (button.length && button.text().trim() !== 'Select') {
                    optionName = button.text().trim();
                }

                options.push({ name: optionName, asin: optAsin });
            }
        });

        if (options.length) {
            dimensions[dimName] = options;
        }
    });
    return dimensions;
}

async function scrapeVariations(variationAsin, dimIndex = 0, prefix = '') {
    if (scraped.has(variationAsin)) return;

    const url = `https://www.amazon.com/dp/${variationAsin}/?th=1&psc=1`;
    const $ = await fetchPage(url);

    const titleElem = $('#productTitle');
    if (!titleElem.length) {
        console.log(`${prefix}[SKIP] ${variationAsin}: Product page not available`);
        scraped.add(variationAsin);
        return;
    }
    const name = titleElem.text().trim();

    // Price extraction
    let price;
    if ($('#outOfStockBuyBox_feature_div').length) {
        price = 'Out of Stock';
    } else {
        const whole = $('.a-price-whole').first().text();
        const fraction = $('.a-price-fraction').first().text();
        price = whole && fraction ? `$${whole}${fraction}` : 'N/A';
    }

    const pageDims = extractDimensions($);

    // End of recursion - collect final data
    if (dimIndex >= dimNames.length || Object.keys(pageDims).length === 0) {
        scraped.add(variationAsin);
        const selections = {};

        for (const dimName of dimNames) {
            const rowIds = [`inline-twister-row-${dimName}_name`, `inline-twister-row-${dimName}`];
            for (const rowId of rowIds) {
                const row = $(`#${rowId}`);
                if (row.length) {
                    const selected = row.find('span[class*="a-button-selected"]');
                    if (selected.length) {
                        const swatch = selected.find('span.swatch-title-text-display');
                        const img = selected.find('img');
                        if (swatch.length) {
                            selections[dimName] = swatch.text().trim();
                        } else if (img.length && img.attr('alt')) {
                            selections[dimName] = img.attr('alt').trim();
                        } else {
                            const text = selected.text().trim();
                            selections[dimName] = text.includes('Select') ? 'N/A' : text;
                        }
                        break;
                    } else {
                        const matchOption = row.find(`li[data-asin="${variationAsin}"]`);
                        if (matchOption.length) {
                            const swatch = matchOption.find('span.swatch-title-text-display');
                            const button = matchOption.find('span.a-button-text');
                            if (swatch.length) {
                                selections[dimName] = swatch.text().trim();
                            } else if (button.length && !button.text().includes('Select')) {
                                selections[dimName] = button.text().trim();
                            }
                            break;
                        }
                    }
                    break;
                }
            }
        }

        const row = [variationAsin, name, ...dimNames.map(dim => selections[dim] || 'N/A'), price];
        results.push(row);

        const selStr = Object.entries(selections).map(([k, v]) => `${k}:${v}`).join(', ');
        console.log(`${prefix}[OK] ${variationAsin}: ${price} | ${selStr}`);
        return;
    }

    // Continue recursion through dimensions
    const currentDim = dimNames[dimIndex];
    if (pageDims[currentDim]) {
        const options = pageDims[currentDim];
        if (dimIndex === 0) {
            console.log(`${prefix}Found ${options.length} ${currentDim} options`);
        }

        for (let i = 0; i < options.length; i++) {
            const option = options[i];
            if (dimIndex === 0) {
                console.log(`${prefix}${currentDim} ${i + 1}/${options.length}: ${option.name}`);
            }
            await scrapeVariations(option.asin, dimIndex + 1, prefix + '    ');
        }
    } else {
        await scrapeVariations(variationAsin, dimIndex + 1, prefix);
    }
}

async function main() {
    // Initial page scrape to find dimensions
    const $ = await fetchPage(targetUrl);
    const dimensions = extractDimensions($);

    console.log(`Found dimensions: ${Object.keys(dimensions).join(', ')}`);
    for (const [dimName, options] of Object.entries(dimensions)) {
        console.log(`  ${dimName}: ${options.length} options`);
    }

    // Setup dimension traversal order and CSV headers
    const priority = ['color', 'size', 'style', 'pattern', 'material', 'fit'];
    dimNames = Object.keys(dimensions).sort((a, b) => {
        const aIdx = priority.indexOf(a.toLowerCase());
        const bIdx = priority.indexOf(b.toLowerCase());
        return (aIdx === -1 ? priority.length : aIdx) - (bIdx === -1 ? priority.length : bIdx);
    });
    headers = ['ASIN', 'Product Name', ...dimNames, 'Price'];

    console.log(`\nDimension traversal order: ${dimNames.join(', ')}`);
    console.log('\nStarting variation crawling...');

    await scrapeVariations(asin);

    // Export to CSV
    const csvWriter = createCsvWriter({
        path: 'productVariations.csv',
        header: headers.map(h => ({ id: h, title: h }))
    });

    // Convert results array to objects
    const records = results.map(row => {
        const obj = {};
        headers.forEach((h, i) => obj[h] = row[i]);
        return obj;
    });

    await csvWriter.writeRecords(records);

    console.log(`\nDone! Scraped ${results.length} unique variations`);
    console.log('Data exported to productVariations.csv');
}

main().catch(console.error);
